src/net/server/message.rs

```diff
@@ -0,0 +1,402 @@
+use octseq::Octets;
+use tokio::time::Instant;
+use tracing::Level;
+use tracing::{enabled, error, info_span, warn};
+
+use crate::base::Message;
+use crate::net::server::buf::BufSource;
+use crate::net::server::metrics::ServerMetrics;
+use crate::net::server::middleware::chain::MiddlewareChain;
+
+use super::service::{CallResult, Service, ServiceError, Transaction};
+use super::util::start_reply;
+
+//------------ Request -------------------------------------------
+
+#[derive(Debug, Copy, Clone)]
+pub struct UdpSpecificTransportContext {
+    pub max_response_size_hint: Option<u16>,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct NonUdpTransportContext {
+    pub idle_timeout: Option<Duration>,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum TransportSpecificContext {
+    Udp(UdpSpecificTransportContext),
+    NonUdp(NonUdpTransportContext),
+}
+
+impl TransportSpecificContext {
+    pub fn is_udp(&self) -> bool {
+        matches!(self, Self::Udp(_))
+    }
+
+    pub fn is_non_udp(&self) -> bool {
+        matches!(self, Self::NonUdp(_))
+    }
+}
+
+/// A DNS message with additional properties describing its context.
+///
+/// DNS messages don't exist in isolation, they are received from somewhere or
+/// created by something. This type wraps a message with additional context
+/// about its origins so that decisions can be taken based not just on the
+/// message itself but also on the circumstances surrounding its creation and
+/// delivery.
+#[derive(Debug)]
+pub struct Request<T> {
+    client_addr: std::net::SocketAddr,
+    received_at: Instant,
+    message: Arc<T>,
+    transport_specific: TransportSpecificContext,
+}
+
+impl<T> Request<T> {
+    pub fn new(
+        client_addr: std::net::SocketAddr,
+        received_at: Instant,
+        message: T,
+        transport_specific: TransportSpecificContext,
+    ) -> Self {
+        Self {
+            client_addr,
+            received_at,
+            message: Arc::new(message),
+            transport_specific,
+        }
+    }
+
+    /// When was this message received?
+    pub fn received_at(&self) -> Instant {
+        self.received_at
+    }
+
+    /// Get the transport specific context
+    pub fn transport(&self) -> &TransportSpecificContext {
+        &self.transport_specific
+    }
+
+    /// From which IP address and port number was this message received?
+    pub fn client_addr(&self) -> std::net::SocketAddr {
+        self.client_addr
+    }
+
+    /// Read access to the inner message
+    pub fn message(&self) -> &T {
+        &self.message
+    }
+}
+
+//--- Clone
+
+impl<T> Clone for Request<T> {
+    fn clone(&self) -> Self {
+        Self {
+            client_addr: self.client_addr,
+            received_at: self.received_at,
+            message: Arc::clone(&self.message),
+            transport_specific: self.transport_specific,
+        }
+    }
+}
+
+//----------- MessageProcessor -----------------------------------------------
+
+pub struct MessageDetails<Buf>
```

Comment:
This is indeed not ideal. `MessageDetails` is just a wrapper to reduce the number of arguments passed around, and is 66% overlap (2/3 fields) with `Request` and those 2 fields are actually copied from `MessageDetails` into `Request` so I can understand the confusion.

`MessageDetails` differs to `Request` in two ways:
1. It contains the buffer which the `MessageProcessor` will use to deserialize the wire message into.
2. `MessageDetails` is used by `MessageProcessor` and relates to receiving of a message, while `Request` is a context wrapper created by `MessageProcessor` to give extra insight to the rest of the processing chain (middleware and service). We don't create `Request` with the rest of the context until after the wire message has been deserialized because it is the server that is asked to add the context and does so bsaed on which type of server it is.

I think this needs improving.

Reply:
I removed `MessageDetails` already by extracting its fields and using them as separate fn arguments for now (resulting in Clippy needing to be silenced with `#[allow(clippy::too_many_arguments)]`), and renamed `MessageProcessor` to `CommonMessageF low`.

This doesn't resolve the code smell but does remove the confusing additional `MessageDetails` type.

I've added a note to the list of things we want to look at going forwards and will resolve this comment as I don't plan to change this further in this PR.